### ISSUE TITLE
rustc built by MIR overflows its stack for crates with very deep ASTs.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#35408]
These examples have been found on crater:
- [ ] [`boiler`](https://crates.io/crates/boiler) ([failure](https://tools.taskcluster.net/task-inspector/#EILyCYnlRbS82RuQCWat4g/0))
- [ ] [`procinfo`](https://crates.io/crates/procinfo) ([failure](https://tools.taskcluster.net/task-inspector/#kEbUuqQvSXO7CwxiiMVKpw/0))
- [ ] [`marksman_escape`](https://crates.io/crates/marksman_escape) ([failure](https://tools.taskcluster.net/task-inspector/#UZZWBDQZTRuPdqn7qx6Icw/0))

These reproduce with `stage2` but not `stage1`, unless `stage0` is passed `-Z orbit`, which means a compiler built by MIR trans will use more stack space than old trans did.

For `procinfo`, there are just over 1000 frames, usually in the `NodeIdAssigner`, my guess is that its AST has a depth 500-700 in places (it's using `nom`, which might cause such extreme nesting).

cc @rust-lang/compiler

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by eddyb** [comment]
May be fixed by #35764, we should do a crater run after that's merged.

**Comment by eddyb** [comment]
Actually, how are we going to do a crater run if we don't have an old trans switch?

I suppose before-after on that PR might suffice (and it would catch other problems wrt drop flags).
I'll start it now then, no point in waiting.

**Comment by nikomatsakis** [comment]
@eddyb  well we can certainly just check the results for `procinfo` by hand, as well

**Comment by eddyb** [comment]
@nikomatsakis My manual check suggests it's fixed, but I want to see it on crater before closing.

**Comment by eddyb** [comment]
[Crater report](https://gist.github.com/eddyb/5013119a94c9758a2f676c1b89e1e29b) shows it didn't actually get fixed.

**Comment by nikomatsakis** [comment]
@eddyb were you able to observe it failing before?

**Comment by eddyb** [comment]
@nikomatsakis Yes, and a stage2 build of #35764 was the first time I saw it work again.

**Comment by eddyb** [comment]
@brson's beta crater runs reveals that @nagisa's [`marksman_escape`](https://github.com/nagisa/marksman_escape/) also overflows the stack (but not for me).
If it *is* the change to remove filling drop from `syntax::ptr::P::map`, we should backport it to beta.
However, crater doesn't agree that the stack overflows have been fixed on master, which is troubling.

**Comment by nagisa** [comment]
Its probably caused by the auto generated trie in
https://github.com/nagisa/marksman_escape/blob/master/src/unescape_named_gen.rs
.

We should really investigate that arbitrary sized initializers do not break
our compiler, likely by removing a great deal of recursion from
visitors'n'stuff.

**Comment by nagisa** [comment]
For all its worth I coud not reproduce the issue with nightly from 08-04, but could reproduce with stage2 build of rustc built from my i128 PR.

**Comment by eddyb** [comment]
I've managed to get the IR from old and MIR trans for `hir::lowering::LoweringContext::lower_expr`, and looking for `alloca`s with no `llvm.lifetime.start` calls:  https://gist.github.com/eddyb/0b9afd3a5cc04fcc35643afb3aa00804

The `[140 x i8]` `alloca`s are `hir::Expr` reduced to `hir::Expr_` (+ 4 unused bytes AFAICT).
Just the `[140 x i8]` and `hir::Expr_` `alloca`s together add up to 9kB, about half of the frame size.
The original `alloca`s (e.g. `hir::lowering::LoweringContext::expr`'s `node` argument) have `StorageLive` and `StorageDead` as needed, but it looks like SROA may be stripping them in some cases.

I need to try to reproduce this outside of `librustc` though.

**Comment by eddyb** [comment]
Something I noticed while looking at #36023 which *may* be relevant: we seem to be emitting `StorageDead` for *all* bindings in *all* arms in the merge of a `match` instead of on each arm.
Not sure if anything better can be done here. Also, the `StorageLive` are in the appropriate arms, so perhaps this has no negative impact at all.

**Comment by nikomatsakis** [comment]
@eddyb do you think that having *more* lifetime-start calls would cause problems? I guess it's possible.

**Comment by brson** [comment]
@eddyb Can this problem be worked around with a bigger stack?

~12 days till this hits stable.

Nominating for more compiler team eyes.

**Comment by danburkert** [comment]
Hi, author of procinfo here.  Is this something that should be working around in procinfo with a recursion_limit annotation?  The nesting is indeed caused by some massive `nom` parsers.  The regression remains with the latest nightly (`rustc 1.13.0-nightly (5531c314a 2016-09-12)`), however there have been nightlies in the recent past which have been fine (`rustc 1.13.0-nightly (499484f56 2016-08-18)`, `cargo 0.13.0-nightly (8f20561 2016-07-16)`).

CC danburkert/procinfo-rs#9
CC danburkert/procinfo-rs#10

**Comment by eddyb** [comment]
@danburkert `#![recursion_limit]` has no effect on the actual compiler stack size, I'm afraid.

**Comment by nrc** [comment]
Discussed at compiler meeting. We think this is a serious regression that must be addressed. However, it seems for those crates affected, there is a simple work around which is setting `RUST_MIN_STACK` to 16MB. So there are two possible solutions: either we set the default `RUST_MIN_STACK` to 16MB (temporarily) or we leave it as is and mention in the release notes and elsewhere that this fix exists. We are not aware of any risk/downside for setting the default `RUST_MIN_STACK`, however, there are some unknowns there.

We decided that we will experiment with the RUST_MIN_STACK minimum now (since time is short). If there are problems or @nikomatsakis objects, we can backout. We should discuss this issue again next week to check if there were problems, and if not backport that PR to beta.

@nagisa will implement the RUST_MIN_STACK change.

(Leaving nominated for further discussion next week)

**Comment by nagisa** [comment]
Something from the meeting that was left unmentioned: We also could be setting RUST_MIN_STACK on buildbots to make sure we don’t regress stack usage infinitely. Something like 2 or 4MiB should be plenty to bootstrap rustc.

**Comment by brson** [comment]
I suspect that the continued failures on crater are due to the taskcluster environment causing the maximum stack size rlimit to be set to something low, and pthread_attr_setstacksize lying about succeeding, but I have not proven it.

**Comment by brson** [comment]
The rlimit stack size seems to only affect the main thread, so I don't think it should be applying here (and I can't get it to affect a spawned thread in a test case).

**Comment by brson** [comment]
Perhaps another thing we could do is instrument std to check whether the stack size is the same as requested and run in through crater, but so far I can't find any documentation that suggests that pthread_attr_setstacksize ever returns success without actually using the requested size.

**Comment by brson** [comment]
Marking this one a regression again because the underlying cause is still aproblem per https://github.com/rust-lang/rust/issues/36778

**Comment by brson** [comment]
@eddyb says they are still waiting for some MIR optimizations

**Comment by brson** [comment]
P-medium because we believe the workaround basically works despite the yet-unexplained weirdness on crater.

**Comment by kamalmarhubi** [comment]
This has regressed in `rustc 1.14.0-nightly (5665bdf3e 2016-11-02)`. It was fine on the last nightly, `rustc 1.14.0-nightly (3f4408347 2016-10-27`.

**Comment by alexcrichton** [comment]
@kamalmarhubi oh thanks for the report! This is a pretty long issue, so could you clarify which crate is failing to compile for you?

**Comment by kamalmarhubi** [comment]
Sorry, procinfo. See https://github.com/danburkert/procinfo-rs/issues/12.

**Comment by kamalmarhubi** [comment]
(It has the dubious distinction of failing to compile on all channels at present.)

**Comment by alexcrichton** [comment]
Ah yep, that'd do it! The changes for that range look like this: https://github.com/rust-lang/rust/compare/3f4408347d2109803edbf53c89c8bce575de4b67...7c69b0d5ae8c404c4847fa214126ebd7c0246277

**Comment by kamalmarhubi** [comment]
Heads up, I got https://github.com/danburkert/procinfo-rs/pull/15 merged with a workaround for this issue in `procinfo`. Passing crater on that crate won't mean it's fixed.

**Comment by brson** [comment]
@rust-lang/compiler @eddyb please retriage this one. When we made it P-medium it was because we thought it was fixed, but it is not.

**Comment by kamalmarhubi** [comment]
Note that it was fixed (or did not manifest on `procinfo`) and then regressed: https://github.com/rust-lang/rust/issues/35408#issuecomment-258281024

**Comment by eddyb** [comment]
@kamalmarhubi What value of `RUST_MIN_STACK` does it start working again at? Default should be `16000000` (~16MB) and IIRC `7000000` (~7MB) used to be enough, maybe we regressed somewhere else (the first dozen frames of a backtrace would be helpful).

**Comment by arielb1** [comment]
> Something I noticed while looking at #36023 which may be relevant: we seem to be emitting `StorageDead` for all bindings in all arms in the merge of a match instead of on each arm.

Sure. The region hierarchy for match-expressions has the bindings destroyed only on the exit block of the match, so for example
```Rust

fn example(ex: Box<Result<[u32; 64], [i32; 64]>>) {
    match *ex {
        Ok(a) => {},
        Err(b) => {}
    }
}
```

Translates to
```
fn example(_1: Box<std::result::Result<[u32; 64], [i32; 64]>>) -> () {
    let mut _0: ();                      // return pointer
    scope 1 {
        let _2: Box<std::result::Result<[u32; 64], [i32; 64]>>; // "ex" in scope 1 at <anon>:3:12: 3:14
        scope 2 {
            let _4: [u32; 64];           // "a" in scope 2 at <anon>:5:12: 5:13
        }
        scope 3 {
            let _5: [i32; 64];           // "b" in scope 3 at <anon>:6:13: 6:14
        }
    }
    let mut _3: ();

    bb0: {
        StorageLive(_2);                 // scope 0 at <anon>:3:12: 3:14
        _2 = _1;                         // scope 0 at <anon>:3:12: 3:14
        switch((*_2)) -> [Ok: bb3, Err: bb4]; // scope 1 at <anon>:5:9: 5:14
    }

    bb1: {
        resume;                          // scope 0 at <anon>:3:1: 8:2
    }

    bb2: {
        drop(_2) -> bb1;                 // scope 0 at <anon>:8:2: 8:2
    }

    bb3: {
        StorageLive(_4);                 // scope 1 at <anon>:5:12: 5:13
        _4 = (((*_2) as Ok).0: [u32; 64]); // scope 1 at <anon>:5:12: 5:13
        _0 = ();                         // scope 2 at <anon>:5:18: 5:20
        goto -> bb5;                     // scope 1 at <anon>:4:5: 7:6
    }

    bb4: {
        StorageLive(_5);                 // scope 1 at <anon>:6:13: 6:14
        _5 = (((*_2) as Err).0: [i32; 64]); // scope 1 at <anon>:6:13: 6:14
        _0 = ();                         // scope 3 at <anon>:6:19: 6:21
        goto -> bb5;                     // scope 1 at <anon>:4:5: 7:6
    }

    bb5: {
        StorageDead(_5);                 // scope 1 at <anon>:7:6: 7:6
        StorageDead(_4);                 // scope 1 at <anon>:7:6: 7:6
        drop(_1) -> [return: bb6, unwind: bb2]; // scope 0 at <anon>:8:2: 8:2
    }

    bb6: {
        drop(_2) -> bb7;                 // scope 0 at <anon>:8:2: 8:2
    }

    bb7: {
        StorageDead(_2);                 // scope 0 at <anon>:8:2: 8:2
        return;                          // scope 1 at <anon>:8:2: 8:2
    }
}
```

Here the live-ranges for `a` and `b` overlap - I suspect that this can lead to LLVM placing them in separate allocas.

**Comment by eddyb** [comment]
@arielb1 Last I looked the numbers seemed to fit the non-overlapping situation, but maybe I wasn't paying enough attention. It'd be great if we can get a win from changing that?

**Comment by nikomatsakis** [comment]
triage: P-high 

We should at least figure out if something changed since the last time we looked, though the solution seems pretty unclear.

**Comment by arielb1** [comment]
All functions in librustc with with stack traces over 4kB:
```
$ objdump -d ./build/x86_64-unknown-linux-gnu/stage0/lib/librustc-62b3e239.so | grep -B10 'sub.*0x[1-9a-f][0-9a-f][0-9a-f][0-9a-f],%rsp' | grep ^000000
00000000000c3260 <_ZN56_$LT$syntax..ast..Expr$u20$as$u20$core..clone..Clone$GT$5clone17h0b9f5bdd0ef90b32E>:
0000000000121a30 <_ZN5rustc9dep_graph6thread4main17h8ba9da102f1e3585E>:
000000000013fe30 <_ZN143_$LT$rustc..hir..lowering..LoweringContext..lower_crate..ItemLowerer$LT$$u27$lcx$C$$u20$$u27$interner$GT$$u20$as$u20$syntax..visit..Visitor$GT$10visit_item17h129c67de0682c23dE>:
00000000001444b0 <_ZN5rustc3hir8lowering15LoweringContext10lower_item17h6c926aaa4dad761aE>:
0000000000147b80 <_ZN5rustc3hir8lowering15LoweringContext10lower_expr17h06dc253ad1c81fa2E>:
000000000017f570 <_ZN5rustc5infer15error_reporting87_$LT$impl$u20$rustc..infer..InferCtxt$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$20report_region_errors17h238676c09050d0a9E>:
0000000000184760 <_ZN5rustc5infer15error_reporting87_$LT$impl$u20$rustc..infer..InferCtxt$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$23report_concrete_failure17hb96357b0dcaa6fcdE>:
00000000001f3df0 <_ZN5rustc6middle8liveness8visit_fn17h2f36d9c5e4b23d1eE>:
000000000022a390 <_ZN5rustc7session6config21rustc_short_optgroups17h1dd178b6bc0868faE>:
000000000022f5e0 <_ZN5rustc7session6config38build_session_options_and_crate_config17hec8089707b9377d4E>:
0000000000239ec0 <_ZN5rustc7session14build_session_17h0115f416cfde5a00E>:
0000000000240620 <_ZN5rustc6traits15error_reporting87_$LT$impl$u20$rustc..infer..InferCtxt$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$22report_selection_error17h7adb94259b4617e9E>:
0000000000262320 <_ZN5rustc6traits6select16SelectionContext17confirm_candidate17h38a1154fd14f35b0E>:
0000000000269350 <_ZN5rustc6traits10specialize20specialization_graph5Graph6insert17h47b0f75b980067a8E>:
```

**Comment by arielb1** [comment]
And in `librustc_mir`:
```
$ objdump -d ./build/x86_64-unknown-linux-gnu/stage0/lib/librustc_mir-62b3e239.so | grep -B10 'sub.*0x[1-9a-f][0-9a-f][0-9a-f][0-9a-f],%rsp' | grep ^000000
0000000000049d40 <_ZN9rustc_mir5build4expr9as_rvalue89_$LT$impl$u20$rustc_mir..build..Builder$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$14expr_as_rvalue17h0dd0e5275ecc6520E>:
000000000004e660 <_ZN9rustc_mir5build4expr9as_rvalue89_$LT$impl$u20$rustc_mir..build..Builder$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$15build_binary_op17he2784c8a7fe205d0E>:
00000000000515c0 <_ZN9rustc_mir5build4expr4into89_$LT$impl$u20$rustc_mir..build..Builder$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$9into_expr17h4593b1066fa67afaE>:
00000000000542f0 <_ZN9rustc_mir5build4expr4stmt89_$LT$impl$u20$rustc_mir..build..Builder$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$9stmt_expr17h8b4bbc53caa2ededE>:
0000000000058630 <_ZN9rustc_mir5build7matches4test89_$LT$impl$u20$rustc_mir..build..Builder$LT$$u27$a$C$$u20$$u27$gcx$C$$u20$$u27$tcx$GT$$GT$12perform_test17h2c3e67e7a182cf74E>:
0000000000075450 <_ZN9rustc_mir7mir_map8BuildMir19build_const_integer17h73d205e0ccc5bfecE>:
0000000000076620 <_ZN126_$LT$rustc_mir..mir_map..BuildMir$LT$$u27$a$C$$u20$$u27$tcx$GT$$u20$as$u20$rustc..hir..intravisit..Visitor$LT$$u27$tcx$GT$$GT$10visit_item17hfdd565e96e1b3b12E>:
0000000000078f30 <_ZN126_$LT$rustc_mir..mir_map..BuildMir$LT$$u27$a$C$$u20$$u27$tcx$GT$$u20$as$u20$rustc..hir..intravisit..Visitor$LT$$u27$tcx$GT$$GT$16visit_trait_item17hed4decf6474376d5E>:
000000000007a350 <_ZN126_$LT$rustc_mir..mir_map..BuildMir$LT$$u27$a$C$$u20$$u27$tcx$GT$$u20$as$u20$rustc..hir..intravisit..Visitor$LT$$u27$tcx$GT$$GT$15visit_impl_item17h595951c1feb85398E>:
000000000007b740 <_ZN126_$LT$rustc_mir..mir_map..BuildMir$LT$$u27$a$C$$u20$$u27$tcx$GT$$u20$as$u20$rustc..hir..intravisit..Visitor$LT$$u27$tcx$GT$$GT$8visit_fn17hbbf343a5ae162072E>:
000000000008a420 <_ZN9rustc_mir9transform10type_check11TypeChecker16check_terminator17h7c541b550bedaad2E>:
000000000008f550 <_ZN110_$LT$rustc_mir..transform..type_check..TypeckMir$u20$as$u20$rustc..mir..transform..MirPass$LT$$u27$tcx$GT$$GT$8run_pass17h90921022119d9ee7E>:
```

**Comment by eddyb** [comment]
@arielb1 Nice! Could that be used on a stable release, prior to MIR trans being on by default?
At least we could see how much worse MIR codegen itself is compared to old trans, in frame sizes.

**Comment by danburkert** [comment]
`procinfo` `0.3.0` is the published version that was affected.  An easy repro is:

```

**Comment by brson** [comment]
cc @eddyb @nikomatsakis any updates? This is still tagged P-high so please circle back to it soon.

**Comment by brson** [comment]
No changes since 11/17. Still P-high. @nikomatsakis @eddyb please re-triage.

**Comment by nikomatsakis** [comment]
triage: P-medium

Still worthy of investigation -- and I hope to do so post-xmas -- but not a burning priority. More of a long-standing issue we need to improve bit by bit.

**Comment by pnkfelix** [comment]
There's been mucho activity investigating stack blowup issues on #40883. @arielb1 believes that the LLVM fixes linked from https://github.com/rust-lang/rust/issues/40883#issuecomment-290997988 will be necessary to resolve these problems for us.

**Comment by arielb1** [comment]
I believe this should be fixed with the fix to #40883.

**Comment by jonas-schievink** [comment]
This can probably be closed now.
